File: src/problem/backjoon/platinum/BJ11003.java
package problem.backjoon.platinum;

import java.io.*;
import java.util.*;

public class BJ11003 {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));

        int[] info = Arrays.stream(br.readLine().split(" ")).mapToInt(i -> Integer.parseInt(i)).toArray();
        StringTokenizer stz = new StringTokenizer(br.readLine());

        Deque<Node> deque = new ArrayDeque<>();

        for (int i=0; i<info[0]; i++) {
            int now = Integer.parseInt(stz.nextToken());

            while (!deque.isEmpty() && deque.getLast().val > now) deque.removeLast();
            deque.addLast(new Node(i, now));

            if (deque.getFirst().idx <= i - info[1]) deque.removeFirst();

            bw.write(deque.getFirst().val + " ");
        }

        bw.flush();
        bw.close();
        br.close();
    }

    static class Node {
        public int idx;
        public int val;

        Node (int i, int v) {
            this.idx = i;
            this.val = v;
        }

    }
}

File: src/problem/platinum/BJ11003.kt
package problem.platinum

import java.io.BufferedReader
import java.io.InputStreamReader
import java.util.*
import kotlin.collections.ArrayDeque

fun main(args: Array<String>) {
    val br = BufferedReader(InputStreamReader(System.`in`))
    val (n, l) = br.readLine().split(" ").map { it.toInt() }
    val stz = StringTokenizer(br.readLine())

    val deq = ArrayDeque<Node>()
    val sb = StringBuilder()

    for (i in 0 until n) {
        val now = stz.nextToken().toInt()

        while (deq.isNotEmpty() && deq.last().value > now) deq.removeLast()
        deq.addLast(Node(i, now))
        if (deq.first().index <= i - l) deq.removeFirst()
        sb.append(deq.first().value).append(" ")

    }
    println(sb.toString())
    br.close()
}

class Node(var index: Int, var value: Int)